# Setup Instructions

Before running this notebook:

1. **Set up the environment** (if not already done):
   ```bash
   ./setup_env.sh
   ```
   Or manually:
   ```bash
   python3 -m venv venv
   source venv/bin/activate
   pip install -r requirements.txt
   python -m ipykernel install --user --name=commodities --display-name="Python (commodities)"
   ```

2. **Select the kernel**: In Jupyter, select the kernel named **"Python (commodities)"** from the kernel selector.


In [3]:
import pandas as pd
import os

# Define the raw data directory
raw_data_dir = 'raw_data'

# List of files to load
files = [
    'COT_FutsOnly_2023.xls',
    'COT_FutsOnly_2024.xls',
    'COT_FutsOnly_2025.xls'
]

# Load and concatenate all files
dataframes = []
for file in files:
    file_path = os.path.join(raw_data_dir, file)
    df = pd.read_excel(file_path)
    print(f"Loaded {file}: {len(df)} rows, {len(df.columns)} columns")
    dataframes.append(df)

# Concatenate vertically
combined_df = pd.concat(dataframes, ignore_index=True)
print(f"\nCombined dataset: {len(combined_df)} rows, {len(combined_df.columns)} columns")

# Display column names to verify Asofdate column
print("\nColumn names:")
print(combined_df.columns.tolist())


Loaded COT_FutsOnly_2023.xls: 12534 rows, 188 columns
Loaded COT_FutsOnly_2024.xls: 13601 rows, 188 columns
Loaded COT_FutsOnly_2025.xls: 9826 rows, 188 columns

Combined dataset: 35961 rows, 188 columns

Column names:
['Market_and_Exchange_Names', 'As_of_Date_In_Form_YYMMDD', 'Report_Date_as_MM_DD_YYYY', 'CFTC_Contract_Market_Code', 'CFTC_Market_Code', 'CFTC_Region_Code', 'CFTC_Commodity_Code', 'Open_Interest_All', 'Prod_Merc_Positions_Long_ALL', 'Prod_Merc_Positions_Short_ALL', 'Swap_Positions_Long_All', 'Swap__Positions_Short_All', 'Swap__Positions_Spread_All', 'M_Money_Positions_Long_ALL', 'M_Money_Positions_Short_ALL', 'M_Money_Positions_Spread_ALL', 'Other_Rept_Positions_Long_ALL', 'Other_Rept_Positions_Short_ALL', 'Other_Rept_Positions_Spread_ALL', 'Tot_Rept_Positions_Long_All', 'Tot_Rept_Positions_Short_All', 'NonRept_Positions_Long_All', 'NonRept_Positions_Short_All', 'Open_Interest_Old', 'Prod_Merc_Positions_Long_Old', 'Prod_Merc_Positions_Short_Old', 'Swap_Positions_Long_Old

In [2]:
# Sort by Asofdate column
# Handle case-insensitive column name matching
date_col = None
for col in combined_df.columns:
    if 'asofdate' in str(col).lower() or 'as_of_date' in str(col).lower():
        date_col = col
        break

if date_col:
    combined_df = combined_df.sort_values(by=date_col)
    print(f"Sorted by '{date_col}' column")
else:
    print("Warning: Could not find Asofdate column. Available columns:")
    print(combined_df.columns.tolist())

# Display first few rows
combined_df.head()


Sorted by 'As_of_Date_In_Form_YYMMDD' column


Market_and_Exchange_Names  \
5424  CHEESE (CASH-SETTLED) - CHICAGO MERCANTILE EXC...   
2889  DOMINION - SOUTH POINT (BASIS) - ICE FUTURES E...   
7615  PJM WESTERN HUB RT PEAK MINI - ICE FUTURES ENE...   
5051      MILK, Class III - CHICAGO MERCANTILE EXCHANGE   
1822        USGC HSFO (PLATTS) - ICE FUTURES ENERGY DIV   

      As_of_Date_In_Form_YYMMDD Report_Date_as_MM_DD_YYYY  \
5424                     230103                2023-01-03   
2889                     230103                2023-01-03   
7615                     230103                2023-01-03   
5051                     230103                2023-01-03   
1822                     230103                2023-01-03   

     CFTC_Contract_Market_Code CFTC_Market_Code  CFTC_Region_Code  \
5424                    063642             CME                  0   
2889                    0233A3             IFED                 1   
7615                    0643DF             IFED                 1   
5051                    052641             CME                  0   
1822                    02141B             IFED                 1   

      CFTC_Commodity_Code  Open_Interest_All  Prod_Merc_Positions_Long_ALL  \
5424                   63              18169                         10764   
2889                   23             782124                        603490   
7615                   64             292626                        149378   
5051                   52              26913                          9097   
1822                   21              33758                         15025   

      Prod_Merc_Positions_Short_ALL  ...  Conc_Gross_LE_4_TDR_Short_Other  \
5424                           8448  ...                              0.0   
2889                         159911  ...                              0.0   
7615                         179712  ...                              0.0   
5051                          10434  ...                              0.0   
1822                          26835  ...                              0.0   

      Conc_Gross_LE_8_TDR_Long_Other  Conc_Gross_LE_8_TDR_Short_Other  \
5424                             0.0                              0.0   
2889                             0.0                              0.0   
7615                             0.0                              0.0   
5051                             0.0                              0.0   
1822                             0.0                              0.0   

      Conc_Net_LE_4_TDR_Long_Other  Conc_Net_LE_4_TDR_Short_Other  \
5424                           0.0                            0.0   
2889                           0.0                            0.0   
7615                           0.0                            0.0   
5051                           0.0                            0.0   
1822                           0.0                            0.0   

      Conc_Net_LE_8_TDR_Long_Other  Conc_Net_LE_8_TDR_Short_Other  \
5424                           0.0                            0.0   
2889                           0.0                            0.0   
7615                           0.0                            0.0   
5051                           0.0                            0.0   
1822                           0.0                            0.0   

                     Contract_Units  CFTC_SubGroup_Code  FutOnly_or_Combined  
5424   (CONTRACTS OF 20,000 POUNDS)                 A25              FutOnly  
2889                     2500 mmbtu                 N13              FutOnly  
7615                           1 MW                 N16              FutOnly  
5051  (CONTRACTS OF 200,000 POUNDS)                 A25              FutOnly  
1822                   1000 barrels                 N10              FutOnly  

[5 rows x 188 columns]

In [3]:
# Display basic info about the dataset
print("Dataset Info:")
print(f"Shape: {combined_df.shape}")
print(f"\nData types:")
print(combined_df.dtypes)
print(f"\nSummary statistics:")
combined_df.describe()


Dataset Info:
Shape: (35961, 188)

Data types:
Market_and_Exchange_Names                object
As_of_Date_In_Form_YYMMDD                 int64
Report_Date_as_MM_DD_YYYY        datetime64[ns]
CFTC_Contract_Market_Code                object
CFTC_Market_Code                         object
                                      ...      
Conc_Net_LE_8_TDR_Long_Other            float64
Conc_Net_LE_8_TDR_Short_Other           float64
Contract_Units                           object
CFTC_SubGroup_Code                       object
FutOnly_or_Combined                      object
Length: 188, dtype: object

Summary statistics:


As_of_Date_In_Form_YYMMDD      Report_Date_as_MM_DD_YYYY  \
count               35961.000000                          35961   
mean               239879.980868  2024-05-24 09:24:41.504963840   
min                230103.000000            2023-01-03 00:00:00   
25%                230919.000000            2023-09-19 00:00:00   
50%                240528.000000            2024-05-28 00:00:00   
75%                250128.000000            2025-01-28 00:00:00   
max                250923.000000            2025-09-23 00:00:00   
std                  7799.441771                            NaN   

       CFTC_Region_Code  CFTC_Commodity_Code  Open_Interest_All  \
count      35961.000000         35961.000000       3.596100e+04   
mean           0.925336            59.982453       1.554763e+05   
min            0.000000             1.000000       7.010000e+02   
25%            1.000000            23.000000       1.940400e+04   
50%            1.000000            64.000000       4.317600e+04   
75%            1.000000            64.000000       1.410260e+05   
max            1.000000           866.000000       7.985639e+06   
std            0.262852            96.445031       4.895365e+05   

       Prod_Merc_Positions_Long_ALL  Prod_Merc_Positions_Short_ALL  \
count                  3.596100e+04                   3.596100e+04   
mean                   7.848271e+04                   6.470114e+04   
min                    0.000000e+00                   0.000000e+00   
25%                    1.085800e+04                   1.010000e+04   
50%                    2.413800e+04                   2.692900e+04   
75%                    6.339100e+04                   6.738800e+04   
max                    4.171781e+06                   2.139422e+06   
std                    2.637057e+05                   1.283268e+05   

       Swap_Positions_Long_All  Swap__Positions_Short_All  \
count             35961.000000               3.596100e+04   
mean              19434.787993               3.872213e+04   
min                   0.000000               0.000000e+00   
25%                 824.000000               7.500000e+02   
50%                4813.000000               4.388000e+03   
75%               15450.000000               1.508900e+04   
max              743658.000000               4.061181e+06   
std               51557.690244               2.318820e+05   

       Swap__Positions_Spread_All  ...  Conc_Net_LE_8_TDR_Long_Old  \
count                35961.000000  ...                35961.000000   
mean                  6931.441506  ...                   59.489152   
min                      0.000000  ...                    0.000000   
25%                    216.000000  ...                   47.300000   
50%                   1366.000000  ...                   61.800000   
75%                   5049.000000  ...                   74.100000   
max                 307042.000000  ...                  100.000000   
std                  22447.169603  ...                   20.054564   

       Conc_Net_LE_8_TDR_Short_Old  Conc_Gross_LE_4_TDR_Long_Other  \
count                 35961.000000                    35961.000000   
mean                     61.906324                        2.198284   
min                       0.000000                        0.000000   
25%                      48.500000                        0.000000   
50%                      63.600000                        0.000000   
75%                      76.700000                        0.000000   
max                     100.000000                      100.000000   
std                      19.764623                       10.457339   

       Conc_Gross_LE_4_TDR_Short_Other  Conc_Gross_LE_8_TDR_Long_Other  \
count                      35961.00000                    35961.000000   
mean                           2.27725                        2.900531   
min                            0.00000                        0.000000   
25%                            0.00000                      

In [5]:
# Display the sorted dataframe
list(combined_df.columns)


['Market_and_Exchange_Names',
 'As_of_Date_In_Form_YYMMDD',
 'Report_Date_as_MM_DD_YYYY',
 'CFTC_Contract_Market_Code',
 'CFTC_Market_Code',
 'CFTC_Region_Code',
 'CFTC_Commodity_Code',
 'Open_Interest_All',
 'Prod_Merc_Positions_Long_ALL',
 'Prod_Merc_Positions_Short_ALL',
 'Swap_Positions_Long_All',
 'Swap__Positions_Short_All',
 'Swap__Positions_Spread_All',
 'M_Money_Positions_Long_ALL',
 'M_Money_Positions_Short_ALL',
 'M_Money_Positions_Spread_ALL',
 'Other_Rept_Positions_Long_ALL',
 'Other_Rept_Positions_Short_ALL',
 'Other_Rept_Positions_Spread_ALL',
 'Tot_Rept_Positions_Long_All',
 'Tot_Rept_Positions_Short_All',
 'NonRept_Positions_Long_All',
 'NonRept_Positions_Short_All',
 'Open_Interest_Old',
 'Prod_Merc_Positions_Long_Old',
 'Prod_Merc_Positions_Short_Old',
 'Swap_Positions_Long_Old',
 'Swap_Positions_Short_Old',
 'Swap_Positions_Spread_Old',
 'M_Money_Positions_Long_Old',
 'M_Money_Positions_Short_Old',
 'M_Money_Positions_Spread_Old',
 'Other_Rept_Positions_Long_Old',
 

## Column Selection

The dataset has 188 columns. Let's organize and select only the essential columns:


**Column Categories:**
- **Identification**: Market, Date, CFTC Codes
- **Positions (All)**: Open Interest, Long/Short positions by trader type
- **Positions (Old/Other)**: Separate reporting categories (usually redundant)
- **Changes**: Week-over-week changes
- **Percentages**: Percent of open interest
- **Traders Count**: Number of traders in each category
- **Concentration**: Top 4/8 trader concentration ratios


In [4]:
# Define essential columns to keep
# Modify this list based on your analysis needs

essential_columns = [
    # Identification columns
    'Market_and_Exchange_Names',
    'As_of_Date_In_Form_YYMMDD',
    'Report_Date_as_MM_DD_YYYY',
    'CFTC_Contract_Market_Code',
    'CFTC_Commodity_Code',
    'Contract_Units',

    # Core position data (All contracts)
    'Open_Interest_All',

    # Producer/Merchant positions
    'Prod_Merc_Positions_Long_ALL',
    'Prod_Merc_Positions_Short_ALL',

    # Swap Dealer positions
    'Swap_Positions_Long_All',
    'Swap__Positions_Short_All',
    'Swap__Positions_Spread_All',

    # Money Manager positions
    'M_Money_Positions_Long_ALL',
    'M_Money_Positions_Short_ALL',
    'M_Money_Positions_Spread_ALL',

    # Other Reportable positions
    'Other_Rept_Positions_Long_ALL',
    'Other_Rept_Positions_Short_ALL',
    'Other_Rept_Positions_Spread_ALL',

    # Total Reportable positions
    'Tot_Rept_Positions_Long_All',
    'Tot_Rept_Positions_Short_All',

    # Non-Reportable positions
    'NonRept_Positions_Long_All',
    'NonRept_Positions_Short_All',

    # Percentages of Open Interest (often more useful than absolute values)
    'Pct_of_OI_Prod_Merc_Long_All',
    'Pct_of_OI_Prod_Merc_Short_All',
    'Pct_of_OI_Swap_Long_All',
    'Pct_of_OI_Swap_Short_All',
    'Pct_of_OI_M_Money_Long_All',
    'Pct_of_OI_M_Money_Short_All',
    'Pct_of_OI_Tot_Rept_Long_All',
    'Pct_of_OI_Tot_Rept_Short_All',
]

# Filter to keep only essential columns
# Check which columns exist (handle any naming variations)
columns_to_keep = [col for col in essential_columns if col in combined_df.columns]

print(f"Keeping {len(columns_to_keep)} out of {len(combined_df.columns)} columns")
print(f"\nColumns to keep:")
for col in columns_to_keep:
    print(f"  ✓ {col}")

# Create filtered dataframe
filtered_df = combined_df[columns_to_keep].copy()

print(f"\nFiltered dataset shape: {filtered_df.shape}")
filtered_df.head()


Keeping 30 out of 188 columns

Columns to keep:
  ✓ Market_and_Exchange_Names
  ✓ As_of_Date_In_Form_YYMMDD
  ✓ Report_Date_as_MM_DD_YYYY
  ✓ CFTC_Contract_Market_Code
  ✓ CFTC_Commodity_Code
  ✓ Contract_Units
  ✓ Open_Interest_All
  ✓ Prod_Merc_Positions_Long_ALL
  ✓ Prod_Merc_Positions_Short_ALL
  ✓ Swap_Positions_Long_All
  ✓ Swap__Positions_Short_All
  ✓ Swap__Positions_Spread_All
  ✓ M_Money_Positions_Long_ALL
  ✓ M_Money_Positions_Short_ALL
  ✓ M_Money_Positions_Spread_ALL
  ✓ Other_Rept_Positions_Long_ALL
  ✓ Other_Rept_Positions_Short_ALL
  ✓ Other_Rept_Positions_Spread_ALL
  ✓ Tot_Rept_Positions_Long_All
  ✓ Tot_Rept_Positions_Short_All
  ✓ NonRept_Positions_Long_All
  ✓ NonRept_Positions_Short_All
  ✓ Pct_of_OI_Prod_Merc_Long_All
  ✓ Pct_of_OI_Prod_Merc_Short_All
  ✓ Pct_of_OI_Swap_Long_All
  ✓ Pct_of_OI_Swap_Short_All
  ✓ Pct_of_OI_M_Money_Long_All
  ✓ Pct_of_OI_M_Money_Short_All
  ✓ Pct_of_OI_Tot_Rept_Long_All
  ✓ Pct_of_OI_Tot_Rept_Short_All

Filtered dataset shape: (3596

Market_and_Exchange_Names  \
5424  CHEESE (CASH-SETTLED) - CHICAGO MERCANTILE EXC...   
2889  DOMINION - SOUTH POINT (BASIS) - ICE FUTURES E...   
7615  PJM WESTERN HUB RT PEAK MINI - ICE FUTURES ENE...   
5051      MILK, Class III - CHICAGO MERCANTILE EXCHANGE   
1822        USGC HSFO (PLATTS) - ICE FUTURES ENERGY DIV   

      As_of_Date_In_Form_YYMMDD Report_Date_as_MM_DD_YYYY  \
5424                     230103                2023-01-03   
2889                     230103                2023-01-03   
7615                     230103                2023-01-03   
5051                     230103                2023-01-03   
1822                     230103                2023-01-03   

     CFTC_Contract_Market_Code  CFTC_Commodity_Code  \
5424                    063642                   63   
2889                    0233A3                   23   
7615                    0643DF                   64   
5051                    052641                   52   
1822                    02141B                   21   

                     Contract_Units  Open_Interest_All  \
5424   (CONTRACTS OF 20,000 POUNDS)              18169   
2889                     2500 mmbtu             782124   
7615                           1 MW             292626   
5051  (CONTRACTS OF 200,000 POUNDS)              26913   
1822                   1000 barrels              33758   

      Prod_Merc_Positions_Long_ALL  Prod_Merc_Positions_Short_ALL  \
5424                         10764                           8448   
2889                        603490                         159911   
7615                        149378                         179712   
5051                          9097                          10434   
1822                         15025                          26835   

      Swap_Positions_Long_All  ...  NonRept_Positions_Long_All  \
5424                     3583  ...                         479   
2889                    63328  ...                       13928   
7615                    43025  ...                         631   
5051                     8190  ...                        2734   
1822                      908  ...                         240   

      NonRept_Positions_Short_All  Pct_of_OI_Prod_Merc_Long_All  \
5424                          576                          59.2   
2889                         8589                          77.2   
7615                          542                          51.0   
5051                         6410                          33.8   
1822                          217                          44.5   

      Pct_of_OI_Prod_Merc_Short_All  Pct_of_OI_Swap_Long_All  \
5424                           46.5                     19.7   
2889                           20.4                      8.1   
7615                           61.4                     14.7   
5051                           38.8                     30.4   
1822                           79.5                      2.7   

      Pct_of_OI_Swap_Short_All  Pct_of_OI_M_Money_Long_All  \
5424                       5.8                         0.0   
2889                      71.2                         6.3   
7615                       3.8                         7.8   
5051                       1.4                         3.0   
1822                       0.6                         0.0   

      Pct_of_OI_M_Money_Short_All  Pct_of_OI_Tot_Rept_Long_All  \
5424                          5.9                         97.4   
2889                          0.4                         98.2   
7615                          2.4                         99.8   
5051                         13.0                         89.8   
1822                          0.0                         99.3   

      Pct_of_OI_Tot_Rept_Short_All  
5424                          96.8  
2889                          98.9  
7615                          99.8  
5051                          76.2  
1822                          99.4  

[5 rows x 30 columns]

## Export to Database

Export the filtered data to SQLite database for use in the visualization app.


In [ ]:
import sqlite3

# Clean column names for SQL (replace spaces and special chars)
filtered_df.columns = [col.replace(' ', '_').replace('(', '').replace(')', '').replace('-', '_')
                      for col in filtered_df.columns]

# Convert date column to proper datetime if needed
if 'As_of_Date_In_Form_YYMMDD' in filtered_df.columns:
    try:
        filtered_df['As_of_Date_In_Form_YYMMDD'] = pd.to_datetime(
            filtered_df['As_of_Date_In_Form_YYMMDD'], format='%y%m%d', errors='coerce'
        )
    except:
        pass

# Create database connection
db_path = 'commodities.db'
conn = sqlite3.connect(db_path)

# Write to database
filtered_df.to_sql('cot_data', conn, if_exists='replace', index=False)

# Create indexes for better query performance
conn.execute('CREATE INDEX IF NOT EXISTS idx_market ON cot_data(Market_and_Exchange_Names)')
conn.execute('CREATE INDEX IF NOT EXISTS idx_date ON cot_data(As_of_Date_In_Form_YYMMDD)')
conn.execute('CREATE INDEX IF NOT EXISTS idx_commodity ON cot_data(CFTC_Commodity_Code)')

conn.commit()
conn.close()

print(f"✅ Data exported successfully to {db_path}")
print(f"   Total rows: {len(filtered_df)}")
print(f"   Total columns: {len(filtered_df.columns)}")
print(f"\nTo run the visualization app, execute:")
print("   streamlit run app.py")


# Testing Yahoo Finance Price Fetching

This section tests the price fetching functionality for commodities.


In [4]:
import yfinance as yf

# Import price fetching modules
from fetch_prices import (
    fetch_historical_prices,
    fetch_commodity_price,
    fetch_multiple_commodities,
    fetch_all_available_commodities
)
from commodity_ticker_mapping import (
    get_ticker_for_commodity,
    normalize_commodity_name,
    get_all_tickers_for_commodity,
    get_all_mapped_commodities
)

print("✓ Price fetching modules imported successfully")


✓ Price fetching modules imported successfully


## Test 1: Check Ticker Mappings

Let's see which commodities from our database can be mapped to Yahoo Finance tickers.


In [5]:
# Get unique commodities from database
import sqlite3

conn = sqlite3.connect('commodities.db')
query = "SELECT DISTINCT Commodity_Name FROM cot_data ORDER BY Commodity_Name"
commodities_df = pd.read_sql_query(query, conn)
conn.close()

all_commodities = commodities_df['Commodity_Name'].tolist()
print(f"Total unique commodities in database: {len(all_commodities)}")

# Check which ones can be mapped
mapped_commodities = []
unmapped_commodities = []
ticker_mapping = {}

for commodity in all_commodities:
    normalized = normalize_commodity_name(commodity)
    ticker = get_ticker_for_commodity(normalized)
    if ticker:
        mapped_commodities.append(commodity)
        ticker_mapping[commodity] = ticker
    else:
        unmapped_commodities.append(commodity)

print(f"\n✓ Commodities with ticker mappings: {len(mapped_commodities)}")
print(f"✗ Commodities without ticker mappings: {len(unmapped_commodities)}")

# Show some mapped commodities
if mapped_commodities:
    print("\nSample mapped commodities:")
    for comm in mapped_commodities[:10]:
        print(f"  {comm} → {ticker_mapping[comm]}")
    if len(mapped_commodities) > 10:
        print(f"  ... and {len(mapped_commodities) - 10} more")


Total unique commodities in database: 421

✓ Commodities with ticker mappings: 75
✗ Commodities without ticker mappings: 346

Sample mapped commodities:
  #2 HEATING OIL- NY HARBOR-ULSD → HO=F
  ARGUS WTI Mid/WTI TRADE MONTH → CL=F
  BLACK SEA WHEAT FINANCIAL → ZW=F
  BRENT CRUDE OIL LAST DAY → CL=F
  BRENT LAST DAY → BZ=F
  BUTTER (CASH SETTLED) → DB=F
  CHEESE (CASH-SETTLED) → DA=F
  CME MILK IV → DC=F
  COCOA → CC=F
  COFFEE C → KC=F
  ... and 65 more


## Test 2: Fetch Prices for a Single Commodity

Let's test fetching prices for a single commodity (Gold).


In [6]:
# Test fetching Gold prices
print("Fetching Gold prices (last 30 days)...")
gold_prices = fetch_commodity_price('GOLD', period='1mo')

if not gold_prices.empty:
    print(f"✓ Successfully fetched {len(gold_prices)} days of data")
    print(f"\nDate range: {gold_prices['DATE'].min()} to {gold_prices['DATE'].max()}")
    print(f"\nLatest prices:")
    print(gold_prices[['DATE', 'OPEN', 'HIGH', 'LOW', 'CLOSE', 'VOLUME']].tail())
else:
    print("✗ No data fetched. This might be due to:")
    print("  - Yahoo Finance API temporarily unavailable")
    print("  - Network connectivity issues")
    print("  - Rate limiting")


Fetching Gold prices (last 30 days)...
✓ Successfully fetched 22 days of data

Date range: 2025-10-06 00:00:00-04:00 to 2025-11-04 00:00:00-05:00

Latest prices:
                        DATE         OPEN         HIGH          LOW  \
17 2025-10-29 00:00:00-04:00  3983.699951  3983.699951  3983.699951   
18 2025-10-30 00:00:00-04:00  3960.000000  4027.199951  3913.699951   
19 2025-10-31 00:00:00-04:00  4034.500000  4034.500000  3976.600098   
20 2025-11-03 00:00:00-05:00  3976.199951  4020.000000  3959.000000   
21 2025-11-04 00:00:00-05:00  4013.699951  4018.000000  3937.100098   

          CLOSE  VOLUME  
17  3983.699951    1344  
18  4001.300049     935  
19  3982.199951     559  
20  4000.300049     559  
21  3941.300049  244620  


## Test 3: Fetch Prices for Multiple Commodities

Let's test fetching prices for multiple commodities.


In [7]:
# Test fetching multiple commodities
test_commodities = ['GOLD', 'SILVER', 'CRUDE OIL', 'NATURAL GAS', 'CORN']

print(f"Fetching prices for {len(test_commodities)} commodities (last 30 days)...")
print("This may take a moment due to rate limiting...\n")

prices_dict = fetch_multiple_commodities(test_commodities, period='1mo', delay=0.5)

print(f"\n{'='*60}")
print(f"Results Summary:")
print(f"{'='*60}")
for commodity, df in prices_dict.items():
    if not df.empty:
        latest_price = df['CLOSE'].iloc[-1] if 'CLOSE' in df.columns else 'N/A'
        print(f"✓ {commodity:20s}: {len(df):4d} days | Latest: ${latest_price}")
    else:
        print(f"✗ {commodity:20s}: No data")


Fetching prices for 5 commodities (last 30 days)...
This may take a moment due to rate limiting...

Fetching prices for: GOLD
  ✓ Successfully fetched 22 days of data
Fetching prices for: SILVER
  ✓ Successfully fetched 22 days of data
Fetching prices for: CRUDE OIL
  ✓ Successfully fetched 22 days of data
Fetching prices for: NATURAL GAS


$NG=F: possibly delisted; no price data found  (period=1mo) (Yahoo error = "No data found, symbol may be delisted")


  Trying ETF ticker UNG as fallback...
  ✓ Successfully fetched 22 days of data
Fetching prices for: CORN
  ✓ Successfully fetched 22 days of data

Results Summary:
✓ GOLD                :   22 days | Latest: $3941.300048828125
✓ SILVER              :   22 days | Latest: $46.89500045776367
✓ CRUDE OIL           :   22 days | Latest: $60.43000030517578
✓ NATURAL GAS         :   22 days | Latest: $13.930000305175781
✓ CORN                :   22 days | Latest: $430.75


## Test 4: Visualize Price Data

If we successfully fetched prices, let's visualize them.


In [8]:
# Visualize price data if available
if prices_dict:
    import plotly.graph_objects as go
    from plotly.subplots import make_subplots

    # Filter to commodities with data
    valid_prices = {k: v for k, v in prices_dict.items() if not v.empty}

    if valid_prices:
        fig = go.Figure()

        for commodity, df in valid_prices.items():
            if 'DATE' in df.columns and 'CLOSE' in df.columns:
                fig.add_trace(go.Scatter(
                    x=df['DATE'],
                    y=df['CLOSE'],
                    mode='lines',
                    name=commodity,
                    line=dict(width=2)
                ))

        fig.update_layout(
            title='Commodity Prices (Last 30 Days)',
            xaxis_title='Date',
            yaxis_title='Price ($)',
            hovermode='x unified',
            height=600,
            showlegend=True
        )

        fig.show()
    else:
        print("No valid price data to visualize")
else:
    print("No price data available for visualization")


## Test 5: Fetch All Available Commodities

Let's see how many of our commodities can be fetched automatically.


In [9]:
# Fetch prices for all commodities that have ticker mappings
# Note: This will take a while due to rate limiting
# Uncomment to run:

# print("Fetching prices for all mappable commodities...")
# print("This will take several minutes due to rate limiting...\n")
#
# all_prices = fetch_all_available_commodities(
#     mapped_commodities[:20],  # Limit to first 20 for testing
#     period='1mo',
#     delay=0.5
# )
#
# print(f"\nSuccessfully fetched prices for {len(all_prices)} commodities")
#
# # Summary statistics
# if all_prices:
#     print("\nSummary:")
#     for commodity, df in all_prices.items():
#         if not df.empty and 'CLOSE' in df.columns:
#             min_price = df['CLOSE'].min()
#             max_price = df['CLOSE'].max()
#             latest_price = df['CLOSE'].iloc[-1]
#             print(f"{commodity}: ${min_price:.2f} - ${max_price:.2f} (Latest: ${latest_price:.2f})")

print("Uncomment the code above to fetch prices for all commodities")


Uncomment the code above to fetch prices for all commodities


## Test 6: Check Ticker Details

Let's see what tickers are available for specific commodities.


In [10]:
# Check ticker details for some commodities
test_names = ['GOLD', 'SILVER', 'CRUDE OIL', 'NATURAL GAS', 'CORN', 'WHEAT']

print("Ticker Details:")
print("=" * 60)
for name in test_names:
    normalized = normalize_commodity_name(name)
    tickers = get_all_tickers_for_commodity(normalized)

    if tickers:
        print(f"\n{name}:")
        print(f"  Futures: {tickers.get('futures', 'N/A')}")
        print(f"  ETF:     {tickers.get('etf', 'N/A')}")
    else:
        print(f"\n{name}: No ticker mapping found")


Ticker Details:

GOLD:
  Futures: GC=F
  ETF:     GLD

SILVER:
  Futures: SI=F
  ETF:     SLV

CRUDE OIL:
  Futures: CL=F
  ETF:     USO

NATURAL GAS:
  Futures: NG=F
  ETF:     UNG

CORN:
  Futures: ZC=F
  ETF:     CORN

WHEAT:
  Futures: ZW=F
  ETF:     WEAT
